New FPGA reference design: ANR

tyoungsc opened the pull request.

tyoungsc committed:
First commit for new example design: adaptive noise reduction (ANR)

tyoungsc committed:
Implemented anr_demo.py (python script to automate the generation of
data and running of FPGA exe)
Enhancements to anr_gendata.py
Implemented writing of output data to oneAPI main file.
Code cleanup

tyoungsc committed:
Implemented a bit more of the skeleton code including input/output
kernels and loopback ANR kernel

tyoungsc committed:
Added class and parsing for ANR config parameters
Added parsing of the config file
More skeleton code

tyoungsc committed:
Add Paul's stencil function and made them work in oneAPI
Implemented vertical filtering
Added skeleton for horizontal filtering
New standalone script for converting from Bayer to RGB

tyoungsc committed:
First version of the design working in emulation and hardware.
Tweaked Validate function
Implemented separate vertical and horizontal kernels
Fixed bug in stencil kernel

tyoungsc committed:
Implemented a 1D vertical window filter to improve area
Removed stencil_kernel.hpp (replaced by vertical_kernel.hpp)
Added operator= to DataBundle
Added different global define options to CMake
Added option to disable global memory interaction
Fixed bug with multiple elements per cycle not being passed to ANR
(defaulted to 1)

tyoungsc committed:
Removed 2 DSPs
Tons of error checking (static and dynamic)
Code cleanup and comments

tyoungsc committed:
Moved stencil kernels to stencil functions (called inside a kernel) to
better align with IP authoring
Created dma_kernels.hpp to clean up main function
Tons more static asserts and error checking
Updated VCXPROJ file
Added file for metaprogramming math (mp_math.hpp)
Comments

tyoungsc committed:
Formatted Paul's files based on Google format (data_bundle and
shift_reg)
Formatting and comments

tyoungsc committed:
Simplified code by removing option for host and device memory; only
support device memory.

tyoungsc committed:
Added more file IO error checking

tyoungsc committed:
Implemented alpha blending by passing original pixel and intensity sigma
downstream

tyoungsc committed:
Optimized version 1:
Created class to represent QFP numbers (using ac_int)
Created python script to generate a exponential (exp()) LUT
Added header files for a exp() and sigma i calculation LUT
Precomputed some more values
Added a check to the ANRParams for bitwidth

tyoungsc committed:
Small changes to make disabling global memory work

tyoungsc committed:
Implemented alpha blending using fixed point math
Allocate memory for sigma i LUT in main so memory can be freed
Updated LUT generation script

tyoungsc committed:
Code cleanup

tyoungsc committed:
clang-format

tyoungsc committed:
Implemented exp() and pow2() lookup tables using metaprogramming.
Includes creating constexpr functions for pow(float, int) and
exp(float). The latter was done using the Taylor series expansion (see
mp_math.hpp).

Changes to qfp.hpp to make QFP::ToFP32() constexpr and therefore allow
the LUTs to be constructed at compile time.

Moved all constants to a separate file (constants.hpp) to cleanup main.cpp.

Implemented a base function for creating a constexpr ROM (translates to
an actual ROM in the FPGA). Is the base class for the exp() and pow2() LUTs.

Had to increase he constexpr step length (-fconstexpr-steps=<value>) in the
CMake file to deal with all this fancy metaprogramming.

Some comments and code cleanup.

tyoungsc committed:
Fixed bug in qfp.hpp where offset was fixed to 31 (only worked for 6
exponent bits)
Created and used qfp_inv_lut.hpp to use a LUT for 1/x
Comments and cleanup

tyoungsc committed:
Fixed issue of defaulting max_cols in SubmitANRKernels
Reduced memory usage in column_stencil.hpp by disabling outer loop
pipelining. This is fine since the latency of the outer loop is ~2
cycles.

tyoungsc committed:
Fixed issue with swapping rows and columns when parsing files
Added output pixel saturation to device code
Cleaned up scripts
Added test data to repo

tyoungsc committed:
First draft of README
Added GUID to samples.json
Tons of comments and some code cleanup

tyoungsc committed:
Added missing files to VCXPROJ
Update top-level CMAKE name

tyoungsc committed:
Updated description in sample.json
removed lut generation python file

tyoungsc committed:
Small README updates and comments

tyoungsc committed:
Switched PixelT to use ac_int instead of unsigned char. Templated on a
compile time switch for number of bits (default to 8)

tyoungsc committed:
Added functions to mp_math for extracting FP32 mantissa and exponent
Changed QFP to have constexpr and non-constexpr versions of both ToFP32
and FromFP32
Saved some RAMs using constexpr checks in BilateralFilter1D
Increased constexpr step size

tyoungsc committed:
Used unsigned short values in the LUTs to reduce RAM utilization.
Used to constexpr From/To FP32 methods in QFP to do this

tyoungsc committed:
Reduced DSP utilization by factoring the components of bilateral
gaussian's.
Reduced DSP utiliation by sending integer pixel values through the
internal pipe rather than floats.
Changed the pow2_lut to a LUT using integer pixel values instead of
QFPs, saving M20Ks.
Removed all warnings with change from INTEL:: to ext::intel.
README updates.

tyoungsc committed:
Comments

tyoungsc committed:
Comments and formatting

tyoungsc committed:
Removed pow2 lut since it idoes not scale well with pixel bitwidth
Fixed issues for using bitwidths > 8 (sizeof(PixelT) doesn't work
properly)
Made 'PixelT' a global type and removed it from all templates

tyoungsc committed:
Comments and code cleanup

tyoungsc committed:
Test change

tyoungsc committed:
Small update to README image

tyoungsc committed:
README updates

tyoungsc committed:
Changed Cmake file for setting Quartus seed
README update

tyoungsc committed:
removed unused ENABLE_USM option in Cmake
Added constexpr step increase to VCXPROJ file

tyoungsc committed:
Switched some constants
Only validate the last set of output

tyoungsc committed:
README update with expected output

tyoungsc committed:
Check less than 0 (not equal) to make it cheaper
Format change on loop

tyoungsc committed:
Fixed issues in shift_reg for using 'auto i' instead of 'int i' to
maintain integral constants
Used ac_int in qfp to reduce resource utilization
Removed zero check in ac_int to save area resources and match RTL.

tyoungsc committed:
Code cleanup

tyoungsc committed:
Added IP_MODE option to improve area estimates compared to IP
Code formatting

tyoungsc committed:
Added minimum type for indexing rows and columns
Improved Saturate by changing the negative float check.

tyoungsc committed:
Comments

tyoungsc committed:
Removed loop coalescing from row_stencil to infer shift register.

tyoungsc committed:
Switching default MAX_COLS to 1920 (HD video quality).

tyoungsc committed:
Added II attributes in all places where expecting II=1
code cleanup
comments

tyoungsc committed:
Move iteration over frames outside of row_stencil and column_stencil (in
preparation for IP authoring) -  moved to multiple kernel enqueues which
should be fine since the frames are large and we are using fast kernel
relaunch.
Cleaned up CMakeLists.txt

tyoungsc committed:
Added IP_MODE back to CMakeLists with a comment
Added missing headers from stencil libs

tyoungsc committed:
Removed explicit clock target from CMake

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/scripts/anr_convert_output.py:
```diff
@@ -0,0 +1,21 @@
+
+  out_file_path = out_dir + "/" + filename_no_ext + ".png"
+  print('Writing output file to {}'.format(out_file_path))
+  Image.fromarray(rgb_data, "RGB").save(out_file_path)
```
There is no mention of this Python script in the README.  Did you intend to include it in the design?  If so, I think you need to explain how and when to use it.  And you'll have to convince me this is a good idea.

tyoungsc replied:
Will chat about Python files offline with you.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/scripts/anr_demo.py:
```diff
@@ -0,0 +1,100 @@
+
+################################################################################
+# TODO: display output nicely
+################################################################################
```
Another .py script not mentioned in the README.  This one has 'TODO' comments.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/scripts/anr_gen_data.py:
```diff
@@ -0,0 +1,77 @@
+# convert to RGB and write out reference output image
+img_filtered_rgb = anr_util.bayer_to_rgb(img_filtered_alpha)
+Image.fromarray(img_filtered_rgb).save(img_out_ref_filename, bits=24)
+################################################################################
```
OK, I'm going to stop tagging all of these - I'm assuming you do NOT intend to include all these .py scripts, and they will be deleted.  If that's not the case let's discuss.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/CMakeLists.txt:
```diff
@@ -0,0 +1,179 @@
+  elseif(FPGA_BOARD MATCHES ".*s10.*")
+    set(SEED 2)
+  elseif(FPGA_BOARD MATCHES ".*agilex.*")
+    set(SEED 3)
```
Do we actually need to specify these seeds to meet performance targets?  Or is this just here as a placeholder in case we need it in the future?  If the latter, consider deleting this.

tyoungsc replied:
Potentially. So I will keep them.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/CMakeLists.txt:
```diff
@@ -0,0 +1,179 @@
+  elseif(FPGA_BOARD MATCHES ".*agilex.*")
+    set(SEED 3)
+  else()
+    set(SEED 4)
```
This seems particularly odd . . .

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/CMakeLists.txt:
```diff
@@ -0,0 +1,179 @@
+  elseif(FPGA_BOARD MATCHES ".*agilex.*")
+    set(CLOCK_TARGET 480MHz)
+  else()
+  set(CLOCK_TARGET 300MHz)
```
Indent looks off

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/CMakeLists.txt:
```diff
@@ -0,0 +1,179 @@
+
+# Allow the user to set the compiler to IP mode. In this mode, the full
+# system checks won't be valid, but the area will be reduced
+# e.g. cmake .. -DIP_MODE=1
```
I think you need more detailed comment here.  It's not just that the system checks won't be valid, I think the resulting exe will be non-functional, correct?  And this should only be used to get an area and fmax estimate for the kernels?

tyoungsc replied:
Deleting until we have IP authoring and this becomes more useful for generating an IP

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/CMakeLists.txt:
```diff
@@ -0,0 +1,179 @@
+set_target_properties(${FPGA_TARGET} PROPERTIES COMPILE_FLAGS "${HARDWARE_COMPILE_FLAGS}")
+set_target_properties(${FPGA_TARGET} PROPERTIES LINK_FLAGS "${HARDWARE_LINK_FLAGS} -reuse-exe=${CMAKE_BINARY_DIR}/${FPGA_TARGET}")
+# The -reuse-exe flag enables rapid recompilation of host-only code changes.
+# See DPC++FPGA/GettingStarted/fast_recompile for details.
```
No simulator target?

tyoungsc replied:
No. MVDR is the only one that does. + simulator is not technically supported...

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/anr.hpp:
```diff
@@ -0,0 +1,397 @@
+#ifndef __ANR_HPP__
+#define __ANR_HPP__
+
```
A comment at the top of this file describing its purpose would be helpful.  As I'm reading through the code it just looks like a bunch of random stuff.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/anr.hpp:
```diff
@@ -0,0 +1,397 @@
+  // the middle pixel index
+  constexpr int kMidIdx = filter_size / 2;
+
+  // static asserts
```
This is not a helpful comment.  How about 'validate template parameters'

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/anr.hpp:
```diff
@@ -0,0 +1,397 @@
+#include "qfp_inv_lut.hpp"
+#include "row_stencil.hpp"
+#include "shift_reg.hpp"
+
```
Shouldn't you include unrolled_loop.hpp?

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/anr.hpp:
```diff
@@ -0,0 +1,397 @@
+
+  // convert unsigned pixels to signed
+  ShiftReg<SignedPixelT, filter_size> buffer_signed;
+  UnrolledLoop<0, filter_size>([&](auto i) {
```
Why <0, filter_size> instead of just <filter_size>?  (you seem to have done this in multiple places)

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/anr.hpp:
```diff
@@ -0,0 +1,397 @@
+      // now that we have the exponential power value ('exp_power'), use the
+      // exponential LUT ('ExpLUT') to lookup the result of exp(-exp_power).
+      // NOTE: when creating the exponential LUT, we stored the values of
+      // exp(-x) = 1/exp(x). This avoids negating the value of 'exp_power'.
```
Good explanation

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/anr.hpp:
```diff
@@ -0,0 +1,397 @@
+// sigma (sig_i) that was forwarded from the vertical kernel.
+//
+template <int filter_size, int filter_size_eff = (filter_size + 1) / 2>
+struct HorizontalFunctor {
```
Do you think it's good practice to put 'Functor' in the name of your functors?  Seems a bit redundant, but perhpas it's helpful?

tyoungsc replied:
I like to know that my class is actually just a functor. I like it, but if you are strong against I would change.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/anr.hpp:
```diff
@@ -0,0 +1,397 @@
+  using IntraPipeT = DataBundle<DataForwardStruct, pixels_per_cycle>;
+  using IntraPipe = ext::intel::pipe<IntraPipeID, IntraPipeT>;
+
+  // static asserts
```
Template parameter validation

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/anr.hpp:
```diff
@@ -0,0 +1,397 @@
+  auto spatial_power = BuildGaussianPowers1D<filter_size_eff>(params.sig_s);
+
+  // the functors for the vertical and horizontal kernels.
+  // You can either use a functor or a lamda here, both work.
```
This comment is a bit confusing, and I don't think we should use second person (you) in comments.  How about 'Functors can be used instead of lambdas for kernel code'

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/anr.hpp:
```diff
@@ -0,0 +1,397 @@
+      constexpr InvLUT inv_lut;
+
+      // Start the column stencil.
+      // It will callback to 'vertical_func' with the additional all of the
```
Delete 'additional'

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/column_stencil.hpp:
```diff
@@ -0,0 +1,211 @@
+using namespace hldutils;
+
+//
+// Generic 1D column (i.e., vertical) stencil.
```
delete the comma

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/column_stencil.hpp:
```diff
@@ -0,0 +1,211 @@
+      // │ r ◄─e ◄─g ◄─────┤ FIFO
+      // ├───┼───┼───┤     └───────────
+      // │ r ◄─e ◄─g ◄─────────────────Input
+      // └───┴───┴───┘
```
I like the ASCII art!

What are g, e, and r?  Should it be b (blue) instead of e?  Or are you indicating that these are shift registers (reg)?  Either way, this is a bit unclear.

tyoungsc replied:
These are generic ascii art that Paul made. "reg" is a label, not the value actually shifting.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/column_stencil.hpp:
```diff
@@ -0,0 +1,211 @@
+        OutPipe::write(out_data);
+      }
+
+      // increment the fifo
```
I don't think 'increment' is an operation that is understood for a fifo.  I think you mean 'increment the FIFO read/write pointers'?

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/data_bundle.hpp:
```diff
@@ -0,0 +1,106 @@
+  DataBundle() {}
+
+  DataBundle(const T op) {
+    // unroll in case this constructor is implemented in an HLS component
```
Are these comments about HLS appropriate in a oneAPI code sample?  Don't we also want to unroll these for oneAPI?

tyoungsc replied:
Leftover from Paul. These will be header files and the comments are obvious. deleting.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/dma_kernels.hpp:
```diff
@@ -0,0 +1,130 @@
+//
+// This kernel provides data into the ANR input pipe from either device memory
+// (disable_device_mem == false) or dummy data to avoid connecting to device
+// memory (disable_device_mem == true). The latter allows use to avoid
```
'allows us(e) to avoid' -> 'avoids'

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/dma_kernels.hpp:
```diff
@@ -0,0 +1,130 @@
+// This kernel provides data into the ANR input pipe from either device memory
+// (disable_device_mem == false) or dummy data to avoid connecting to device
+// memory (disable_device_mem == true). The latter allows use to avoid
+// having a device memory interconnect in our kernel system for testing the IP.
```
'testing the IP' needs further explanation here

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/dma_kernels.hpp:
```diff
@@ -0,0 +1,130 @@
+
+  // the number of iterations is the number of total pixels (rows*cols)
+  // divided by the number of pixels per cycle
+  const int iterations = cols * rows / pixels_per_cycle;
```
Do we need to worry about the case where this is not an integer?

tyoungsc replied:
I have a runtime check to ensure `cols` is a multiple of `pixels_per_cycle`.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/dma_kernels.hpp:
```diff
@@ -0,0 +1,130 @@
+      h.single_task<KernelId>([=]() [[intel::kernel_args_restrict]] {
+        device_ptr<T> in(in_ptr);
+
+        [[intel::loop_coalesce(2)]]
```
A comment explaining this loop property and/or referring readers tot he loop_coalesce tutorial would be nice here

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/dma_kernels.hpp:
```diff
@@ -0,0 +1,130 @@
+  // divided by the number of pixels per cycle
+  const int iterations = cols * rows / pixels_per_cycle;
+
+  if constexpr (!disable_device_mem) {
```
Could you move this if inside the loop, to reduce the amount of duplicated code?

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/README.md:
```diff
@@ -0,0 +1,208 @@
+### Quantized Floating-Point (QFP)
+Floating-point values consist of a sign bit, an exponent, and a mantissa. In this design, we take [32-bit single-precision](https://en.wikipedia.org/wiki/Single-precision_floating-point_format) floating values and convert them to quantized floating-point (QFP) values which use less bits. All of the QFPs in this design have 10 bits total, but use a different number for the exponent and mantissa. T ...
+
+### Using Libraries
```
This section is not really about using libraries, it's just listing some header files.  Maybe rename this to 'Reusable Header Files'?

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/README.md:
```diff
@@ -0,0 +1,208 @@
+Floating-point values consist of a sign bit, an exponent, and a mantissa. In this design, we take [32-bit single-precision](https://en.wikipedia.org/wiki/Single-precision_floating-point_format) floating values and convert them to quantized floating-point (QFP) values which use less bits. All of the QFPs in this design have 10 bits total, but use a different number for the exponent and mantissa. T ...
+
+### Using Libraries
+In this design, we use the following generic libraries:
```
Again, reusable/generic header files instead of libraries, just to avoid the heavily overloaded term 'libraries'?

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/main.cpp:
```diff
@@ -0,0 +1,430 @@
+class OutputKernelID;
+
+//
+// Run the ANR algorithm on the FPGA
```
Shouldn't this say 'on the target device', as it could be the emulator device (or in theory some non-FPGA target)?

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/main.cpp:
```diff
@@ -0,0 +1,430 @@
+      SubmitOutputDMA<OutputKernelID, PixelT, ANROutPipe, kPixelsPerCycle,
+                      kDisableDeviceMem>(q, out_ptr, rows, cols, frames);
+
+  // launch all ANR kernel
```
kernels (plural)

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/mp_math.hpp:
```diff
@@ -0,0 +1,157 @@
+#ifndef __MP_MATH__
+#define __MP_MATH__
+
```
Please add a description of the purpose / contents of this file in a comment at the top of the file

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/qfp.hpp:
```diff
@@ -0,0 +1,201 @@
+#ifndef __QFP_HPP__
+#define __QFP_HPP__
+
```
Add file description comment

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/qfp.hpp:
```diff
@@ -0,0 +1,201 @@
+
+  // the number of mantissa bits for the QFP
+  static constexpr unsigned qfp_mantissa_bits =
+      qfp_total_bits - qfp_exponent_bits - is_signed;
```
Is it safe to assume that a bool will always have a value of 0 or 1?

tyoungsc replied:
Yes. By the C++ spec.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/qfp.hpp:
```diff
@@ -0,0 +1,201 @@
+
+    // get the most significant qfp_mantissa_bits from the float's mantissa
+    auto qfp_mantissa =
+        (fp32_mantissa >> mantissa_bit_diff) & qfp_mantissa_mask;
```
Should we be rounding the mantissa instead of truncating?  If you've decided it's not worth the extra area to do proper rounding, it's probably worth a comment to that effect here.

tyoungsc replied:
RTL truncates. What type of comment were you thinking?

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/rom_base.hpp:
```diff
@@ -0,0 +1,68 @@
+#ifndef __ROM_BASE_HPP__
+#define __ROM_BASE_HPP__
+
```
File description comment please

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/row_stencil.hpp:
```diff
@@ -0,0 +1,151 @@
+#ifndef __ROW_STENCIL_HPP__
+#define __ROW_STENCIL_HPP__
+
```
File description comment

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/row_stencil.hpp:
```diff
@@ -0,0 +1,151 @@
+  constexpr int kShiftRegSize = filter_size + parallel_cols - 1;
+  constexpr IndexT kColThreshLow = kPaddingPixels;
+
+  // static asserts
```
template parameter validation

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/row_stencil.hpp:
```diff
@@ -0,0 +1,151 @@
+  }
+
+  // There is a bug where using loop_coalese here prevents the shift register
+  // ('shifty_pixels') from being inferred as shift register.
```
Given the bug, do you not need to manually coalesce the loops?  If not, then why ask the compiler to coalesce them for you?

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/shift_reg.hpp:
```diff
@@ -0,0 +1,141 @@
+#ifndef __SHIFT_REG_HPP__
+#define __SHIFT_REG_HPP__
+
```
File description comment

mdbtucker reviewed (commented):
You seem to have at least one binary file (rilly_input.png - is that rilly how you spell your dog's name?) that is not used or mentioned in the README.  I think if you're going to include a file you need to describe it.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/dma_kernels.hpp:
```diff
@@ -0,0 +1,130 @@
+#ifndef __DMA_KERNELS_HPP__
+#define __DMA_KERNELS_HPP__
+
```
These kernels seem to serve the same purpose as the FakeIOPipes used in several other code samples.  Why not reuse those classes rather than re-inveting the wheel here?  Is there some special functionality in the DMA classes not provided by FakeIOPipes?

tyoungsc committed:
Addressing Mike's reviews
- Added header comments to files that contain more than 1 function/class
- Removed unused scripts and images (kept them locally)
- Removed IP_MODE and DISABLE_GLOBAL_MEM from Cmake file and source
  code; will use this in the future with IP authoring

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/column_stencil.hpp:
```diff
@@ -1,6 +1,8 @@
 #ifndef __COLUMN_STENCIL_HPP__
 #define __COLUMN_STENCIL_HPP__
 
+// TODO:
```
This seems like you added it by mistake?

tyoungsc replied:
Fixed. Grepped for TODO and got rid of all.

mdbtucker reviewed (approved):
After you fix that one 'TODO' comment, please close this PR and create a PR into the main repo development branch.

tyoungsc committed:
Removed TODO

tyoungsc committed:
Removed TODO
